### DISCUSSION TITLE
Path aliases not working with wdio + cucumber

### DISCUSSION THREAD

**Discussion by llwch** [webdriverio/webdriverio#10648] [Q&A]
Hi. 
I have test framework with @wdio/cucumber-framework and want to use path aliases, but see error during test runs: 
`[0-0] 2023-06-30T15:54:43.132Z ERROR @wdio/runner: Error: Cannot find package '@pages/login.page.js' imported from {some path}\wdio_test\features\step-definitions\steps.ts`

Invalid path: import LoginPage from "@pages/login.page.js"; but vscode see no error and can go to the definition.

Also tried to add "tsconfig-paths" package, same error if tsconfig is set up as module in wdio config. 
Is it a bug, or I missed something?

ts-config:

> {
>   "ts-node": {
>     "require": ["tsconfig-paths/register"],
>     "transpileOnly": true,
>     "files": true,
>     "compilerOptions": {}
>   },
>   "compilerOptions": {
>     "module": "Node16",
>     "target": "es2021",
>     "rootDir": "./",
>     "baseUrl": "./",
>     "noEmit": true,
>     "allowImportingTsExtensions": true,
>     "types": [
>       "node",
>       "@wdio/globals/types",
>       "expect-webdriverio",
>       "@wdio/cucumber-framework"
>     ],
>     "paths": {
>       "@pages/*": ["features/pageobjects/*"]
>     }
>   }
> }

Wdio config: 

> import type { Options } from "@wdio/types";
> 
> export const config: Options.Testrunner = {
>   runner: "local",
>   autoCompileOpts: {
>     autoCompile: true,
>     tsNodeOpts: {
>       project: "./tsconfig.json",
>       transpileOnly: true,
>     },
>   },
>   specs: ["../features/login.feature"],
>   exclude: [],
>   maxInstances: 10,
>   capabilities: [
>     {
>       browserName: "chrome",
>     },
>   ],
>   logLevel: "info",
>   bail: 0,
>   baseUrl: "http://localhost",
>   waitforTimeout: 10000,
>   connectionRetryTimeout: 120000,
>   connectionRetryCount: 3,
>   services: ["chromedriver"],
>   framework: "cucumber",
>   reporters: ["spec"],
>   cucumberOpts: {
>     require: ["./features/step-definitions/steps.ts"],
>     backtrace: false,
>     requireModule: [],
>     dryRun: false,
>     failFast: false,
>     snippets: true,
>     source: true,
>     strict: false,
>     tagExpression: "",
>     timeout: 60000,
>     ignoreUndefinedDefinitions: false,
>   },
> };

Versions: 
Node: v18.15.0

**Comment by portyu9** [comment]
Hi @llwch 

There are two separate issues in that configuration.

First, your path mapping doesn't match the import:

```json
"paths": {
    "@pages/": ["features/pageobjects/"]
}
```

`@pages/` is not a wildcard mapping for `@pages/login.page.js`. It should be:

```json
{
    "compilerOptions": {
        "baseUrl": ".",
        "paths": {
            "@pages/*": ["features/pageobjects/*"]
        }
    }
}
```

Then:

```ts
import LoginPage from '@pages/login.page.js'
```

can match the alias.

Second, the configuration in the question uses the old `ts-node` / `autoCompileOpts` setup. Current WebdriverIO uses `tsx` to execute TypeScript and automatically reads the `tsconfig.json`; `tsx` supports `tsconfig.json` path aliases.

So on a current WDIO version I would:

1. fix the alias to `@pages/*`;
2. install/use `tsx` as documented by WDIO;
3. remove the obsolete `autoCompileOpts.tsNodeOpts` setup;
4. keep a separate `tsc --noEmit` step for type checking.

VS Code being able to jump to the file only proves that the TypeScript language service understands the alias; the runtime resolver must understand it too.

I reviewed the full posted `tsconfig`, WDIO config, error, and Node version. There were no replies or attachments.

Please mark this answer as accepted if it helped, thank you.
